pkg/vm/engine/tae/stl/containers/strvec.go

```diff
@@ -127,68 +137,171 @@ func (vec *StrVector[T]) String() string {
+
+		// 1. Remove data from vec.area
+		vec.area.RangeDelete(int(oldOff), int(oldLen))
+		// 2. Update vdata in vec.vdata
+		newVdata[0] = byte(nlen)
+		copy(newVdata[1:1+nlen], val)
+		vec.vdata.Update(i, newVdata)
+		// 3. Adjust offset
+		vec.adjustOffsetLen(i+1, -int(oldLen))
+		return
+	} else if oldVdata.IsSmall() && nlen > types.VarlenaInlineSize {
+		// If the old is small and the new is not
+		var offset int
+		min, max := vec.getAreaRange(0, i)
+		if min == max {
+			offset = 0
+		} else {
+			offset = max
+		}
+		tail := vec.area.Slice()[offset:]
+		val = append(val, tail...)
+		vec.area.RangeDelete(offset, vec.area.Length()-offset)
+		vec.area.AppendMany(val...)
+		newVdata.SetOffsetLen(uint32(offset), uint32(nlen))
+		vec.vdata.Update(i, newVdata)
+		vec.adjustOffsetLen(i+1, nlen)
+		return
+	} else {
+		// If both the old and new are not small and not equal
+
+		offset := int(oldOff + oldLen)
+		tail := vec.area.Slice()[offset:]
+		val = append(val, tail...)
+		vec.area.RangeDelete(int(oldOff), vec.area.Length()-int(oldOff))
+		vec.area.AppendMany(val...)
+		newVdata.SetOffsetLen(oldOff, uint32(nlen))
+		vec.vdata.Update(i, newVdata)
+		vec.adjustOffsetLen(i+1, nlen-int(oldLen))
 	}
 }
 
 func (vec *StrVector[T]) Delete(i int) (deleted T) {
-	s := vec.offsets.Get(i)
-	l := vec.lengths.Get(i)
-
-	deleted = any(vec.data.Slice()[s : s+l]).(T)
-	vec.data.RangeDelete(int(s), int(l))
-	vec.offsets.Delete(i)
-	vec.lengths.Delete(i)
-	for j := i; j < vec.Length(); j++ {
-		old := vec.offsets.Get(j)
-		vec.offsets.Update(j, old-l)
+	deleted = vec.Get(i)
+	vec.RangeDelete(i, 1)
+	return
+}
+
+func (vec *StrVector[T]) getAreaRange(offset, length int) (min, max int) {
+	var pos int
+	slice := vec.vdata.Slice()
+	end := offset + length
+	for pos = offset; pos < end; pos++ {
+		if slice[pos].IsSmall() {
+			continue
+		}
+		min32, _ := slice[pos].OffsetLen()
+		min = int(min32)
+		break
+	}
+	// no data stored in area
+	if pos == end {
+		return
+	}
+	minPos := pos
+	for pos = end - 1; pos >= minPos; pos-- {
+		if slice[pos].IsSmall() {
+			continue
+		}
+		max32, maxLen := slice[pos].OffsetLen()
+		max = int(max32 + maxLen)
+		break
 	}
 	return
 }
 
+func (vec *StrVector[T]) rangeDeleteNoArea(offset, length int) {
+	vec.vdata.RangeDelete(offset, length)
+}
+
+func (vec *StrVector[T]) adjustOffsetLen(from int, delta int) {
+	pos := from
+	slice := vec.vdata.Slice()
+	areaLen := uint32(vec.area.Length())
+	var newOff uint32
+	for pos < vec.Length() {
+		if slice[pos].IsSmall() {
+			pos++
+			continue
+		}
+		oldOff, oldLen := slice[pos].OffsetLen()
+		if delta >= 0 {
+			newOff = oldOff + uint32(delta)
+		} else {
+			newOff = oldOff - uint32(-delta)
+		}
+		slice[pos].SetOffsetLen(newOff, oldLen)
+		if newOff+oldLen == areaLen {
+			break
+		}
+		pos++
+	}
+}
+
 func (vec *StrVector[T]) RangeDelete(offset, length int) {
-	for i := offset + length - 1; i >= offset; i-- {
-		vec.Delete(i)
+	if vec.area.Length() == 0 {
```

Comment:
Can we just not do delete at all?   Just don't touch the area and do not need to adjust offset len.

Does this range delete really help reduce memory?   Why do it?

Reply:
There is some tradeoff. 

The current implementation is not efficient enough at point deletion, since we also adjust the area and need to adjust offset. 
An optional implementation is not to touch the area and only change the data. In this case, when the vector is continuously updated, deleted, and inserted, the size of the area will continue to grow.

I prefer the optional implementation and change it in the next pr by @LeftHandCold. It should also be a local change